### PR TITLE
[Relay][transform] Clean up pass.h

### PR DESCRIPTION
**Pull request by zhiics**
This is the C++ side of item 3 of #3202.

- Remove pass.h
    - Some expr->expr passes, like `ForwardRewrite` and `Bind`, are only invoked on an expression as helper functions in a `FunctionPass` to perform a certain optimization on a relay `Function`. Therefore,  I kept the old expr->expr API as well.
    - Some passes, such as `FreeVars` etc, are not transformation but analysis. I added an `analysis.h` file for them.
    - Theoretically, `analysis` should be also baked into the pass infra. For example, we could register some attributes to preserve validation and other information for an expression/module. Other passes then can directly get the preserved analysis results without specifically invoking these passes.

- Removed the frontend ir_pass.py file and put the functions in either transform or analysis.

- Corresponding changes are made in the following positions for cleanup:
    - [x] C++ files that uses `pass.h`
    - [x] Python unit test
    - [x] Python frontends
    - [x] Python tutorials
    - [x] VTA

### CODE REVIEW

**Comment by tqchen** [comment]
@zhiics I hope to push this change asap, can you please rebase and work on it? We need to update the python side as well(and update the testcases)

**Comment by zhiics** [comment]
sorry. I got really busy recently. I will rebase and push the work I have so far today. I planed to only remove the pass.h in this PR and remove it python ir_pass.xx in a follow-up because I am afraid that it make break things.

**Comment by tqchen** [comment]
ping @zhiics is this PR superseded by others?

**Comment by zhiics** [comment]
@tqchen yeah, I want to have separate PRs to make review easier as there are many files will be modified. I will update this PR after #3430 is merged.

**Comment by zhiics** [comment]
@tqchen @jroesch @wweic @icemelon9 @MarisaKirisame @antinucleon @yzhliu @kevinthesun 

This PR is now ready to review.

File: python/tvm/relay/frontend/tensorflow.py [hunk 1]
```diff
@@ -237,7 +238,7 @@ def _infer_out_shapes(inputs, params):
 
 def _infer_shape(node, params=None):
     """A method to get the output shape of an intermediate node in the relay graph."""
-    out_type = ir_pass.infer_type(node)
+    out_type = _transform.OptimizeOnExpr(node, _transform.InferType())
```
Comment (tqchen) [inline python/tvm/relay/frontend/tensorflow.py]: I don’t think we want to introduce OptimizeOnExpr helper. Instead, we force everything into module to module and manually

Comment (zhiics) [reply python/tvm/relay/frontend/tensorflow.py]: yeah, this helper function internally doesn't module->module opts and returns the entry_func of the updated module. It is mainly used by the python unit tests and cases where only infer type is needed.

I can probably remove them and add it in the corresponding python unit tests. How do you think?

Comment (tqchen) [reply python/tvm/relay/frontend/tensorflow.py]: That sounds good, we do not want to encourage the use of this API

**Pull request merged**